### PR TITLE
Rewrite of UI

### LINKED ISSUE #191: Darkmode
it would be nice, to have a dark mode, based on Browser / System-Enviroment:

https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-color-scheme

### LINKED ISSUE #203: Dark terminal "breaks" ANSI colors
`ansi-bright-white-fg` has a value of `#151515`, which is a rather black color. Ignoring the confusing aspect of bright-white being a rather dark black, after #196 got merged, anything using it is pretty much unreadable, as can be seen in the screenshot below:
image

### LINKED ISSUE #229: Hide UI elements you are not allowed to change/see
# now:
Bildschirmfoto zu 2021-06-18 23-42-08
# expect:
Bildschirmfoto zu 2021-06-18 23-43-34

### and gray-out with a hint on mouse-hover-over to login, on things like "restart" button i.e.  ...

### LINKED ISSUE #246: RFC: Some Web-UI ideas
I would like to help bringing the UI to the next level. The UX can be approved a bit IMO and the development setup would love some care as well.

Some ideas:
- Start to convert Javascript Code to Typescript
- Update libraries
- Use Vite instead of Webpack
- Think if UI should stay with React or use something else (maybe Vue)

Maybe a design update would be nice as well 😉

Peek 2021-07-17 13-34

### LINKED ISSUE #266: Add descriptions (minimal docs) to ui
# Problem
There are currently quite some options inside the settings without any description. Only place you "could" eventually find descriptions for these settings is in the documentation.

## Repo settings
Screenshot from 2021-08-18 00-39-42

For example `Project Visibility` Internal vs Private could be unclear to users.

## Registries

Add some info box what registries / secrets are and how they effect your CI.

Screenshot from 2021-08-18 00-52-31

For example for secrets:

```
Secrets are exposed to your pipeline steps and plugins as uppercase environment variables and 
can therefore be referenced in the commands section of your pipeline.
A secret with the name `test` would be accessible as `TEST` inside your CI step. 
Have a look to the [docs](...) for more information.
```

# Suggestion

To keep things simple and to not need to find the correct place in the docs, I would suggest to add some kind of description explaining each option to all possible places where settings are used. 

Example for `Project Settings > Protected`:

Screenshot from 2021-08-18 00-39-42

### PR DESCRIPTION
**Pull request by anbraten**
closes #246 
closes #229 
closes #266 
closes #203
closes #191 

# Idea
Rewrite of the UI using Typescript, Vue3, Windicss (similar to Tailwind) and Vite. The design should  be close to the current one with some changes:
- latest pipeline in a sidebar on the right
- secrets and registry as part of the repo-settings (secrets and registry entries shouldn't be used as much so they can be "hidden" under settings IMO)
- start page shows list of active repositories with button to enable / add new ones (currently you see all repositories and in most cases you only add new repositories once in a while)

## TODO
<details>

- authentication
  - [x] login
  - [x] logout
  - [x] login error page
  - [x] handle token errors and re-login (an expired token will be removed by server, so no need to handle it by hand)
- own repo list
  - [x] show active repos
  - [x] show button to activate more repos 
- add new repo
  - [x] list inactive & active repos 
- repo overview
  - [x] show builds
  - [x] show current build status (badge)
  - [x] link vcs
  - [x] show list of active branches
- build details
  - [x] show jobs
  - [x] show job logs
  - [x] show details (since, duration)
  - [x] link commit
  - [x] build actions (cancel / re-start pipeline)
  - [x] approve / decline pipeline
  - [x] ~~show artifacts and download links~~
- repo settings
  - [x] secrets
  - [x] registry
  - [x] delete / deactivate repo
  - [x] general settings
- user settings
  - [x] show own token / cli setup
- layout
  - build-feed (right sidebar)
    - [x] open / close
    - [x] improve item view
    - [x] listen to updates
    - [x] ~~add search to filter by repo, commit message~~
  - navbar
    - [x] show amount of currently active & pending pipelines (click on it to open / close sidebar)
- permissions
  - [x] show login button if not authenticated
  - [x] don't show hidden repos (redirect if access)
  - [x] don't show buttons / views if missing permissions 
    - [x] settings
    - [x] hide active jobs & sidebar if not authenticated
    - [x] build actions (restart, cancel, ...)
- ci (add some new ci jobs)
  - [x] check format with prettier
  - [x] check linting
  - [x] check types
- misc
  - [x] link docs
  - [x] show current version
  - [x] icons
  - [x] confirm dialogs
  - [x] add toast for every save / await call
  - [x] support emojis in commit-messages
  - [x] handle errors and toast them
  - [x] add some eslint config
  - [x] allow to set docs url again
  - [x] fix tabs
  - [x] check color contrasts
  - [x]  running gif need transparent background
  - [x] increment time of running builds
  - [x] show no builds in sidebar if empty
  - [x] only show "own" repos in repo-add and repo-list (repos you have at least push access to) #476 
  - [x] add search for repo list
</details>
 
## Before merging
This PR includes some changes introduced by following PRs. So they have to be merged before this one gets merged.
- [x] #251
- [x] #248
- [x] #434 
- [x] #437 
- [x] #436 

## Follow Ups
- remove `WithDocs`
- remove template execution for `index.html` => simply serve static file

## Screenshots 
<details>

Peek 2021-10-26 22-34

Screenshot from 2021-09-19 00-17-24
Screenshot from 2021-09-19 00-16-46
Screenshot from 2021-09-19 00-16-26

Screenshot from 2021-09-19 21-52-55
</details>

### CODE REVIEW

**Comment by 6543** [comment]
Is there a comparsion why we should use vite not weebpack & what are jsx files??

**Comment by anbraten** [comment]
> Is there a comparsion why we should use vite not webpack & what are jsx files??

Vite brings some nice speed improvements while developing compared to webpack.
Additionally it is quite simple to setup and maintain as it is supporting some nice features like typescript, jsx, css-postprocessing, ... out of the box.

JSX files are js files which include UI (html combined with js). [React docs](https://reactjs.org/docs/introducing-jsx.html
)

Most UI files of this project are actually `.jsx` files, but named `.js` which is supported by some fallback mechanism. 

[Vite docs](https://vitejs.dev/guide/why.html)

[Vite Webpack comparison](https://dev.to/vbanditv/vite-a-short-overview-4p17)

**Comment by 6543** [comment]
hey from the screenshots it looks quite ready already.

sorry that I had not much time to look into your pulls lately :/

two things:
A. I think it looks quite ready and  we should consider to move it in asap. esp. the backend code got some refactors I do need to have some thoughts ;)
B. Can you add some frontend TS/Vue linter into it? e.g. https://eslint.vuejs.org/ ...

If you have time I would be happy to chat with you via discord pm

**Comment by anbraten** [comment]
> A. I think it looks quite ready and we should consider to move it in asap. esp. the backend code got some refactors I do need to have some thoughts ;)

Plan was to rewrite the current state with little to no new features and that is "almost" ready. Missing some parts like toast notifications, error handling and the public read-only access to repos. 
I think for the backend changes we should create separate PRs. Some of the changes of this PR are already covered by #251 and #248 

> B. Can you add some frontend TS/Vue linter into it? e.g. https://eslint.vuejs.org/ ...

Sure. I already had that on my list, but haven't searched for a nice config yet.

**Comment by davidak** [comment]
The preview looks very good.

Is it created with a dark mode in mind? That would be very nice.

**Comment by anbraten** [comment]
> The preview looks very good.
> 
> Is it created with a dark mode in mind? That would be very nice.

@davidak To get the basic rewrite done as quick as possible I won't add it directly, but the winidicss (used styling framework) supports dark-mode and I plan to continue to work on such new features after that.

**Comment by anbraten** [comment]
@6543 Aparat from some minor changes like eslint-config, some ci-tasks, pagination edge cases, ... the ui rewrite should be almost finished now. Would be nice if you could try it out and provide me some feedback :wink:

**Comment by 6543** [comment]
nice!!!!

**Comment by 6543** [comment]
can we use https://www.npmjs.com/package/eslint-config-airbnb-typescript (eslint-config-airbnb-base) for linting?

File: cmd/drone-server/flags.go [hunk 1]
```diff
@@ -95,12 +101,6 @@ var flags = []cli.Flag{
 		Usage:  "file path for the drone config",
 		Value:  ".drone.yml",
 	},
-	cli.StringFlag{
```
Comment (6543) [inline cmd/drone-server/flags.go:98]: I think we can promote documentation url somewhere in UI?

Comment (anbraten) [reply cmd/drone-server/flags.go:98]: I removed the flag as I do not see a reason an administrator would need to change that url, but it is visible in the UI. The "new" place I defined which url the docs are located at is: `server/web/config.go`.

**Comment by anbraten** [comment]
> can we use https://www.npmjs.com/package/eslint-config-airbnb-typescript (eslint-config-airbnb-base) for linting?

I have added airbnb config and some other plugins as well. Now I only have to resolved some errors 😂 
Screenshot from 2021-09-13 22-29-17

**Comment by alexef** [comment]
would this filter for the default branch make it in the UI rewrite 🙏 https://github.com/woodpecker-ci/woodpecker/pull/205?

**Comment by anbraten** [comment]
> would this filter for the default branch make it in the UI rewrite  #205?

What do you think about introducing some tabs. It seems like some other CI systems are doing it this way. My suggestion would be to add tabs called: `Activity`, `Branches`, `Pull Requests`, `Deployments`. Somehow like this sample I found from another CI system. This would allow you to go to branches and then to your default branch.

Screenshot from 2021-09-18 17-52-08

File: cmd/drone-server/flags.go (hunk 1 above)
Comment (alexef) [reply cmd/drone-server/flags.go]: our use case is linking out to company internal docs around woodpecker and ci, rather than the public ones

**Comment by alexef** [comment]
> This would allow you to go to branches and then to your default branch.

works for us, despite the extra clicks, thanks!

File: server/router/router.go [hunk 2]
```diff
@@ -187,8 +187,6 @@ func Load(serveHTTP func(w http.ResponseWriter, r *http.Request), middleware ...
 		debugger.GET("/pprof/trace", debug.TraceHandler())
 	}
 
-	e.GET("/api/web-config.js", web.WebConfig)
```
Comment (anbraten) [inline server/router/router.go]: @6543 This was needed for the new UI and works as a replacement for the old template injection.

Comment (6543) [reply server/router/router.go]: @anbraten sorry for that one

**Comment by 6543** [comment]
Bildschirmfoto zu 2021-10-20 16-55-30

I think the background (outside of the bird) should be transparent ... I'll have a look into it

**Comment by davidak** [comment]
The bird also looks very pixelated. Replace with better version if possible. Maybe we need a graphic designer at some point to create good assets.

**Comment by anbraten** [comment]
New woodpecker:

Peek 2021-10-21 14-05

File: server/api/hook.go [hunk 3]
```diff
@@ -275,7 +275,7 @@ func PostHook(c *gin.Context) {
 
 	defer func() {
 		for _, item := range buildItems {
-			uri := fmt.Sprintf("%s/%s/%d", server.Config.Server.Host, repo.FullName, build.Number)
+			uri := fmt.Sprintf("%s/%s/build/%d", server.Config.Server.Host, repo.FullName, build.Number)
```
[the remark this answers is not in the document]
Comment (anbraten) [reply server/api/hook.go:279]: done

**Review by 6543 — APPROVED** [review]
could not find any issue in the code and it looks amazing.

I think there are 1-2... bugs/issues still hidden somewhere but that can be fixed in followups.
This do really bring things forward and fix a lot :)

**Comment by davidak** [comment]
@6543 one check failed https://wp.laszlo.cloud/woodpecker-ci/woodpecker/1079/14

**Comment by davidak** [comment]
Great work. Thank you @6543!

**Comment by 6543** [comment]
forward thanks to @anbraten ;)

**Comment by davidak** [comment]
@6543 ah, sure. you only merged it

**Pull request merged**